contracts/SPVStore.sol

```diff
@@ -3,105 +3,105 @@ pragma solidity 0.4.25;
-    /// @dev            This supports ONLY WITNESS INPUTS AND OUTPUTS
-    /// @param _tx      Raw bytes tx
-    /// @return         Transaction id, little endian
-    function parseAndStoreTransaction(bytes _tx) public returns (bytes32) {
-
-        // If invalid prefix, bubble up error
-        if (!validatePrefix(_tx)) { return; }
-
-        uint8 _numInputs = _tx.extractNumInputs();
-        uint8 _numOutputs = _tx.extractNumOutputs();
-
-        // inputs and outputs
-        // 4 bytes version, 2 bytes flag
-        // 2 bytes numIn + numOut
-        // 41 bytes per input, variable per output
-        uint preWitnessLen = 4 + 2 + 2 + (_numInputs * 41);
+    struct TxOut {
+        uint64 value;               // 8 byte value
+        OutputTypes outputType;
+        bytes payload;              // pubkey hash, script hash, or OP_RETURN data
+    }
 
-        TxIn[] memory _inputs = new TxIn[](_numInputs);
-        TxOut[] memory _outputs = new TxOut[](_numOutputs);
+    struct Transaction {
+        bytes32 txid;               // 32 byte tx id, little endian
+        uint32 locktime;            // 4 byte locktime
+        uint8 numInputs;            // number tx inputs
+        uint8 numOutputs;           // number tx outputs
+        mapping (uint8 => TxIn) inputs;
+        mapping (uint8 => TxOut) outputs;
+    }
 
-        // Parse inputs
-        for (uint8 i = 0; i < _numInputs; i++) {
+    struct Header {
+        bytes32 digest;             // 32 byte little endian digest
+        uint32 version;             // 4 byte version
+        bytes32 prevHash;          // 32 byte previous block hash
+        bytes32 merkleRoot;         // 32 byte tx root
+        uint32 timestamp;           // 4 byte timestamp
+        uint256 target;             // 4 byte nBits == 32 byte integer
+        uint32 nonce;               // 4 byte nonce
+    }
 
-            _inputs[i] = parseInput(_tx.extractInputAtIndex(i));
+    mapping(bytes32 => Transaction) public transactions;    // Transactions
+    mapping(bytes32 => Header) public headers;              // Parsed headers
 
-            // If invalid outpoint, bubble up error
-            if (_inputs[i].outpoint == bytes32(0)) { return; }
-        }
+    /// @notice         Parses, a tx, valides its inclusdion in the block, stores to the mapping
+    /// @param _txid    The transaction hash
+    /// @param _digest  The header hash
+    /// @param _proof   The raw byte proof (concatenated LE hashes)
+    /// @param _index   The index of the leaf
+    /// @return         true if fully valid, false otherwise
+    function validate( bytes32 _txid, bytes32 _digest, bytes _proof, uint _index) public returns (bool) {
 
-        // Parse outputs
-        // Track the number of bytes we process so that we can build the txid later
-        for (i = 0; i < _numOutputs; i++) {
+        // Return false if invalid proof
+        if (!_txid.prove(_digest, headers[_digest].merkleRoot, _proof, _index)) { return false; }
```

Comment:
not sure I like this syntax here, because we're not operating on the `_txid`,

prefer `!prove(_txid, _digest, ...`

Reply:
lolwut